### DISCUSSION TITLE
Simulation with nlmixr2

### DISCUSSION THREAD

**Discussion by lirui0321** [nlmixr2/nlmixr2#283] [Q&A] [answered]
Hello Matt and John,

I have two questions about simulations. 

1.  once I establish a nlmixr2 model, is there an easy way to simulate with only fixed effect. I understand that I can fix all eta values to zero before using simulate() function, but just wonder if there is a quick way to turn off both inter-sub variability and residual errors.   

2. Is there an easy way to simulate individuals in the data set but with a different dosing regimen (instead of randomly generating new individuals)?

Thank you!

**Comment by mattfidler** [comment, marked as answer]
> once I establish a nlmixr2 model, is there an easy way to simulate with only fixed effect. I understand that I can fix all eta values to zero before using simulate() function, but just wonder if there is a quick way to turn off both inter-sub variability and residual errors.

I believe you can use `predict()` on the model and it will do this for you.

There is also the `zeroRe()` functions to create a model that does the same thing as well.

**Reply by lirui0321** [reply]
Many thanks, Matt!

Could you help me take a look the error I posted under another Q&A? It works with the example model with FOCEI, but not FO. And it does not work with my customized model with either FO or FOCEI.

**Reply by mattfidler** [reply]
I have started looking.

**Comment by mattfidler** [comment]
> Is there an easy way to simulate individuals in the data set but with a different dosing regimen (instead of randomly generating new individuals)?

Yes, simply create a new event table for each subject, and get all the parameters from the model to simulate every individual with their estimated parameters:

``` r
library(nlmixr2)
#> Loading required package: nlmixr2data

one.cmt <- function() {
  ini({
    ## You may label each parameter with a comment
    tka <- 0.45 # Log Ka
    tcl <- log(c(0, 2.7, 100)) # Log Cl
    ## This works with interactive models
    ## You may also label the preceding line with label("label text")
    tv <- 3.45; label("log V")
    ## the label("Label name") works with all models
    eta.ka ~ 0.6
    eta.cl ~ 0.3
    eta.v ~ 0.1
    add.sd <- 0.7
  })
  model({
    ka <- exp(tka + eta.ka)
    cl <- exp(tcl + eta.cl)
    v <- exp(tv + eta.v)
    linCmt() ~ add(add.sd)
  })
}

fit <- nlmixr(one.cmt, theo_sd, est="focei",
              control=list(print=0))
#> ℹ parameter labels from comments are typically ignored in non-interactive mode
#> ℹ Need to run with the source intact to parse comments
#> rxode2 3.0.2.9000 using 8 threads (see ?getRxThreads)
#>   no cache: create with `rxCreateCache()`
#> calculating covariance matrix
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00 
#> done
#> → Calculating residuals/tables
#> ✔ done
#> → compress origData in nlmixr2 object, save 5952
#> → compress parHistData in nlmixr2 object, save 6120


# Create a data frame with the parameters of every individual
pars <- fit$eta
for (i in names(fit$theta)) {
  pars[[i]] <- fit$theta[[i]]
}

# Create a new simulation dataset
e <- et(amt=320, ii=8, until=48) %>%
  et(seq(0, 48, 0.1)) %>%
  et(id=1:12) # 12 individuals in fit

# Solve with individual parameter estimates from the fit
# and the new simulation dataset
s <- rxSolve(one.cmt, params=pars, events=e)
#> ℹ parameter labels from comments are typically ignored in non-interactive mode
#> ℹ Need to run with the source intact to parse comments
#> Warning: multi-subject simulation without without 'omega'

plot(s, ipredSim)
```

![](https://i.imgur.com/jmkk1pC.png)

<sup>Created on 2024-11-15 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by lirui0321** [reply]
Many thanks, Matt!
